atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampBoundStore.java

```diff
@@ -194,4 +240,57 @@ private Column makeColumn(long ts) {
     private static ByteBuffer getRowName() {
         return ByteBuffer.wrap(PtBytes.toBytes(ROW_AND_COLUMN_NAME));
     }
+
+    private String replaceBracesWithStringFormatSpecifier(String msg) {
+        return msg.replaceAll("\\{\\}", "%s");
+    }
+
+    @VisibleForTesting
+    long getId() {
+        return id;
+    }
+
+    private static final class IdAndTimestamp {
+        private final Optional<Long> id;
+        private final long timestamp;
+
+        /*
+         * Very much of an edge case, but there is theoretically a risk of a wrong interpretation, if the byte
+         * representation of a long in the old format incidentally happens to match a string with one underscore
+         * somewhere in it. For example,
+         *   PtBytes.toString(PtBytes.toBytes(3557616313682636848L)); // returns "1_100000"
+         */
+        private IdAndTimestamp(byte[] values) {
+            String stringValue = PtBytes.toString(values);
+            Matcher columnStringMatcher = TIMESTAMP_FORMAT_PATTERN.matcher(stringValue);
+            if (columnStringMatcher.matches()) {
+                String[] entries = stringValue.split("_");
+                this.id = Optional.of(Long.parseLong(entries[0]));
+                this.timestamp = Long.parseLong(entries[1]);
+            } else {
+                this.id = Optional.empty();
+                this.timestamp = PtBytes.toLong(values);
+            }
+        }
+
+        private IdAndTimestamp(Column column) {
+            this(column.getValue());
+        }
+
+        private IdAndTimestamp(CASResult result) {
+            this(result.getCurrent_values().get(0).getValue());
```

Comment:
Can we always assume that there's only one value? If so, this might be clearer:
`this(Iterables.getOnlyElement(result.getCurrent_values()).getValue());`